## Part 3: Ethics and Optimization

### Task 2: Troubleshooting Challenge

In [1]:
import tensorflow as tf
from tensorflow.keras import datasets, layers, models

print("=== Buggy Code Example ===")
try:
    buggy_model = tf.keras.Sequential([
        layers.Conv2D(32, (3,3), input_shape=(28,28), activation='relu'),
        layers.Flatten(),
        layers.Dense(10)
    ])
    buggy_model.compile(optimizer='adam', loss='mean_squared_error', metrics=['accuracy'])
    print("Model compiled successfully (unexpected).")
except Exception as e:
    print("Error in buggy model:", e)

=== Buggy Code Example ===
Error in buggy model: Input 0 of layer "conv2d" is incompatible with the layer: expected min_ndim=4, found ndim=3. Full shape received: (None, 28, 28)


C:\Users\DIANA\anaconda3\lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [2]:
# Load and preprocess MNIST dataset
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0
x_train = x_train.reshape(-1, 28, 28, 1)
x_test = x_test.reshape(-1, 28, 28, 1)

In [3]:
# Define corrected CNN model
model = models.Sequential([
    layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])
model

<Sequential name=sequential_1, built=True>

In [4]:
# Compile with correct loss and optimizer
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
model.compile

<bound method Trainer.compile of <Sequential name=sequential_1, built=True>>

In [5]:
# Train model
print("\n=== Training Fixed Model ===")
history = model.fit(x_train, y_train, epochs=5, validation_data=(x_test, y_test), verbose=2)


=== Training Fixed Model ===
Epoch 1/5
1875/1875 - 56s - 30ms/step - accuracy: 0.9576 - loss: 0.1400 - val_accuracy: 0.9839 - val_loss: 0.0482
Epoch 2/5
1875/1875 - 54s - 29ms/step - accuracy: 0.9857 - loss: 0.0466 - val_accuracy: 0.9875 - val_loss: 0.0383
Epoch 3/5
1875/1875 - 50s - 27ms/step - accuracy: 0.9895 - loss: 0.0342 - val_accuracy: 0.9896 - val_loss: 0.0327
Epoch 4/5
1875/1875 - 55s - 29ms/step - accuracy: 0.9927 - loss: 0.0232 - val_accuracy: 0.9889 - val_loss: 0.0325
Epoch 5/5
1875/1875 - 85s - 46ms/step - accuracy: 0.9937 - loss: 0.0190 - val_accuracy: 0.9914 - val_loss: 0.0274


In [6]:
# Evaluate model
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print(f"\n Model successfully fixed and trained.")
print(f"Test Accuracy: {test_acc:.4f}")

313/313 - 3s - 11ms/step - accuracy: 0.9914 - loss: 0.0274

 Model successfully fixed and trained.
Test Accuracy: 0.9914


##  Troubleshooting Challenge; Debugging TensorFlow Model

**Problem Summary:**  
The initial Convolutional Neural Network (CNN) model failed due to shape mismatches and incorrect loss configuration in TensorFlow. These issues prevented successful model compilation and training.

**Identified Issues:**  
1. The input shape was defined as (28, 28) instead of (28, 28, 1), causing a dimension mismatch
2. The model incorrectly used mean squared error as the loss function instead of a classification-appropriate loss.
3. The output layer lacked a softmax activation, preventing proper probability distribution for class predictions.

**Fixes Implemented:**  
- Corrected the input dimension to include the single color channel (28, 28, 1).
- Replaced the loss function with sparse categorical crossentropy, suitable for multi-class classification.
- Added MaxPooling layers to enhance feature extraction and reduce overfitting.
- Included a softmax activation in the final layer to produce class probabilities. 

**Result:**  
After debugging and retraining, the model compiled successfully and achieved the following results on the MNIST test set:

Test Accuracy: 99.14%

Test Loss: 0.0274

The fixes resolved all configuration errors, and the optimized CNN achieved excellent performance, demonstrating effective troubleshooting and TensorFlow debugging skills.
